### DISCUSSION TITLE
Difference between f-string and concatenation in hybrid properties

### DISCUSSION THREAD

**Discussion by MattSom** [sqlalchemy/sqlalchemy#7186] [Usage Questions] [answered]
I have a hybrid property like this:

```python
@hybrid_property
def display_name(self):
    return f"{self.file_name}.{self.extension_id}"
```

My application uses this value on the UI as one of the columns of a table html element, which can be sorted. However sorting on this field I get the following error:

`AttributeError
AttributeError: 'str' object has no attribute 'asc'`

The last part of the traceback:

```python
File "/home/project_path/lib/python3.9/site-packages/sqlalchemy/sql/operators.py", line 1532, in asc_op
def desc_op(a):
    return a.desc()
 
 
def asc_op(a):
    return a.asc()
 
 
def nulls_first_op(a):
    return a.nulls_first()
 
AttributeError: 'str' object has no attribute 'asc'
```

But when I change my hybrid property to this, it works:

```python
@hybrid_property
def display_name(self):
   return self.file_name + "." + self.extension_id
```

Obviously my whole UI sorting logic is obscure here, but this change makes it all good. I'm very confused what's the difference in the output.

**Comment by CaselIT** [comment, marked as answer]
Hi,

`f"{self.file_name}.{self.extension_id}"` returns a string in all cases. `self.file_name + "." + self.extension_id` returns the sum of the 3 elements. When they all happen to be string these are the same, but if they are not strings but columns the returned object is the sum operation of the tree columns, that for string columns is transformed in a concatenation.

**Reply by MattSom** [reply]
Hi,

Thanks for the explanation. However it is still not clear why should this throw this problem. Most of the cases the concatenation returns a string value as well, so why the `AttributeError: 'str' object has no attribute 'asc'` error is not thrown in that case?

**Reply by CaselIT** [reply]
An [hybrid property](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html) is used both on the orm instances and on the orm class when creating a query to database.
When you create a query to the db you cannot return a string, but you need to return a column like object, so that sqlalchemy can propertly render it and/or apply additional operations to it, like `asc`

I would suggest placing a breakpoint (or print) in the property to see how that's used in your code.

**Reply by MattSom** [reply]
I totally understand now. With the f-string the column (on which the sorting query should happen) are not identifiable anymore. With the concatenation however, `SQLAlchemy` sees that "Okay, so you want me to use _this_ and _this_ column."

Thanks for the clarification : )

**Reply by CaselIT** [reply]
Yes, more or less. It's not a just a problem of identification.
When summing columns (or doing most other operations) you don't get a string back, but a sqlalchemy column element, that can then be further used by sqlalchemy:
```py
In [2]: import sqlalchemy as sa
In [3]: a,b = sa.column('a'), sa.column('b')
In [4]: a+'.'+b
Out[4]: <sqlalchemy.sql.elements.BinaryExpression object at 0x00000204338D3370>
In [5]: str(a+'.'+b)
Out[5]: 'a || :a_1 || b'
In [6]: f'{a}.{b}'
Out[6]: 'a.b'
In [7]: type(f'{a}.{b}')
Out[7]: str
```
